creating_geometric_models/triangulate_point_cloud_and_clean_up.py

```diff
@@ -0,0 +1,101 @@
+from sklearn.decomposition import PCA
+import numpy as np
+from scipy.spatial import Delaunay
+import trimesh
+from Face import Face
+import argparse
+import os
+
+parser = argparse.ArgumentParser(
+    description=(
+        "triangulate a point cloud and recursively clean the triangulation based on"
+        " edge-length, area and aspect ratio criteria"
+    )
+)
+parser.add_argument("input_file", help="input file (3 columns ascii x y z)")
+parser.add_argument(
+    "--edge_length_threshold",
+    nargs=1,
+    type=float,
+    help="border triangles with edge length larger than the threshold will be removed",
+)
+parser.add_argument(
+    "--area_threshold",
+    nargs=1,
+    type=float,
+    help="border triangles with area lower than the threshold will be removed",
+)
+parser.add_argument(
+    "--aspect_ratio_threshold",
+    nargs=1,
+    type=float,
+    default=[7.0],
+    help="border triangles with aspect ratio larger than the threshold will be removed",
+)
+
+args = parser.parse_args()
+
+
+xyz = np.loadtxt(args.input_file)
+# Perform PCA to get principal axes
+pca = PCA(n_components=2)
```

Comment:
This part I don't fully get. Why are you running a PCA here?
Don't you typically want to triangulate e.g. a surface mesh in (x,y) or (x,z) or (y,z) coordinates?

Reply:
The example I'm showing is the one of a "normal fault" dipping 60 degrees (angle with the horizontal plane) and with a strike (azimuth from North) of 50 degrees.
I assume the best is to find 2 coordinates u and v which align with the along dip and along strike coordinates of the mean planar fault best approximating the point cloud, and to run the Delaunay using this new coordinate system. That's what I would expect I get from the PCA.
I agree that this might be not needed for most fault: I just tested with xy on this example fault and it looks also reasonable,. The PCA allows automatizing this choice of coordinates to use for triangulating.

Reply:
Ah, I didn't think of faults for some reason. Then the method makes sense but I am not sure whether it is really robust. But as you should check the resulting mesh manually anyway, I wouldn't worry too much about it.
After PCA the first two components should explain most of the variance, which makes sense for meshing.